app/services/events/stores/aggregated_clickhouse_store.rb

```diff
@@ -390,16 +394,82 @@ def sum_date_breakdown
       end
 
       def weighted_sum(initial_value: 0)
-        # TODO(pre-aggregation): Implement
+        result = connection_with_retry do |connection|
+          query = Events::Stores::AggregatedClickhouse::WeightedSumQuery.new(store: self)
+
+          sql = ActiveRecord::Base.sanitize_sql_for_conditions(
+            [
+              sanitize_colon(query.query),
+              {
+                from_datetime:,
+                to_datetime: to_datetime.ceil,
+                decimal_scale: DECIMAL_SCALE,
+                initial_value: initial_value || 0
+              }
+            ]
+          )
+
+          connection.select_one(sql)
+        end
+
+        result["aggregation"]
       end
 
       def grouped_weighted_sum(initial_values: [])
-        # TODO(pre-aggregation): Implement
+        connection_with_retry do |connection|
+          query = AggregatedClickhouse::WeightedSumQuery.new(store: self)
+
+          # NOTE: build the list of initial values for each groups
+          #       from the events in the period
+          formated_initial_values = grouped_count.map do |group|
+            value = 0
+            previous_group = initial_values.find { |g| g[:groups] == group[:groups] }
+            value = previous_group[:value] if previous_group
+            {groups: group[:groups], value:}
+          end
+
+          # NOTE: add the initial values for groups that are not in the events
```

Comment:
Does it make sense to add the initial values if there's no event for those groups in the period ? 🤔

Reply:
In case of a `recurring`  billable metric, the units from the previous period will be kept and considered as consumed for the entire period.